core/src/main/java/org/apache/iceberg/jdbc/JdbcClientPool.java

```diff
@@ -21,17 +21,39 @@
 import java.sql.Connection;
 import java.sql.DriverManager;
 import java.sql.SQLException;
-import java.sql.SQLNonTransientConnectionException;
+import java.sql.SQLTransientException;
+import java.util.Arrays;
 import java.util.Map;
 import java.util.Properties;
+import java.util.Set;
+import java.util.stream.Collectors;
 import org.apache.iceberg.CatalogProperties;
 import org.apache.iceberg.ClientPoolImpl;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableSet;
+import org.apache.iceberg.relocated.com.google.common.collect.Sets;
 
 public class JdbcClientPool extends ClientPoolImpl<Connection, SQLException> {
 
+  /**
+   * The following are common retryable SQLSTATE error codes which are generic across vendors.
+   *
+   * <ul>
+   *   <li>08000: Generic Connection Exception
+   *   <li>08003: Connection does not exist
+   *   <li>08006: Connection failure
+   *   <li>08007: Transaction resolution unknown
+   * </ul>
+   *
+   * See https://en.wikipedia.org/wiki/SQLSTATE for more details.
+   */
+  static final Set<String> COMMON_RETRYABLE_CONNECTION_SQL_STATES =
```

Comment:
Maybe worth to add `40001` and `40P01` state to the list:
- `40001` is the code used by most databases to identify deadlocks
- `40P01` is a PostgreSQL specific code for deadlocks as well (see https://www.postgresql.org/docs/13/errcodes-appendix.html)

Deadlocks are a candidate to retry imho.

Reply:
Sure it seems like 40001 is typically a transaction rollback prompted by a database's deadlock detector kicking in and essentially killing one of the operations.

I think I'll add 40001 since it can be safely retried and is a common status code across vendors.

I'm a bit more hesitant to add database vendor specific codes (like 40P01) to this default set, since it sets a precedent to be a dumping ground for many states. I'd advocate we keep this default set to be a common set, and users pass in custom codes for other cases.

Reply:
`40P01` can be provided by the user, if using PostgreSQL. So, OK to add `40001` state only.